doc/ranger.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man 5.01 (Pod::Simple 3.43)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "RANGER 1"
.TH RANGER 1 ranger-1.9.4 2025-04-09 "ranger manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
ranger \- visual file manager
.SH SYNOPSIS
.IX Header "SYNOPSIS"
\&\fBranger\fR [\fB\-\-version\fR] [\fB\-\-help\fR] [\fB\-\-debug\fR] [\fB\-\-clean\fR]
[\fB\-\-cachedir\fR=\fIdirectory\fR] [\fB\-\-confdir\fR=\fIdirectory\fR]
[\fB\-\-datadir\fR=\fIdirectory\fR] [\fB\-\-copy\-config\fR=\fIwhich\fR]
[\fB\-\-choosefile\fR=\fItarget\fR] [\fB\-\-choosefiles\fR=\fItarget\fR]
[\fB\-\-choosedir\fR=\fItarget\fR] [\fB\-\-selectfile\fR=\fIfilepath\fR]
[\fB\-\-show\-only\-dirs\fR]
[\fB\-\-list\-unused\-keys\fR] [\fB\-\-list\-tagged\-files\fR=\fItag\fR]
[\fB\-\-profile\fR] [\fB\-\-cmd\fR=\fIcommand\fR] [\fIpath ...\fR]
.SH DESCRIPTION
.IX Header "DESCRIPTION"
ranger is a console file manager with VI key bindings.
.SH RESOURCES
.IX Header "RESOURCES"
This \fImanual\fR contains instructions on how to use and configure ranger.
.PP
Inside \fIranger\fR, you can press \fI?\fR for a list of key bindings, commands or
settings.
.PP
The \fIREADME\fR contains install instructions.
.PP
The file \fIHACKING.md\fR contains guidelines for code modification.
.PP
The directory \fIdoc/configs\fR contains configuration files.  They are usually
installed to \fI/usr/share/doc/ranger/config\fR and can be obtained with ranger's
\&\-\-copy\-config option.
.PP
The directory \fIexamples\fR contains reference implementations for ranger
plugins, sample configuration files and some programs for integrating ranger
with other software.  They are usually installed to
\&\fI/usr/share/doc/ranger/examples\fR.
.PP
The man page of \fBrifle\fR\|(1) describes the functions of the file opener
.PP
The section \fILINKS\fR of this man page contains further resources.
.SH "POSITIONAL ARGUMENTS"
.IX Header "POSITIONAL ARGUMENTS"
.IP "\fIpath ...\fR" 14
.IX Item "path ..."
Each path will be opened in a tab and if the path is a file it will be selected.
Omitting this is equivalent to providing the current directory.
.SH OPTIONS
.IX Header "OPTIONS"
.IP "\fB\-d\fR, \fB\-\-debug\fR" 14
.IX Item "-d, --debug"
Activate the debug mode: Whenever an error occurs, ranger will exit and print a
full traceback.  The default behavior is to merely print the name of the
exception in the statusbar/log and try to keep running.
.IP "\fB\-c\fR, \fB\-\-clean\fR" 14
.IX Item "-c, --clean"
Activate the clean mode:  ranger will not access or create any configuration
files nor will it leave any traces on your system.  This is useful when your
configuration is broken, when you want to avoid clutter, etc.
.IP \fB\-\-cachedir\fR=\fIdir\fR 14
.IX Item "--cachedir=dir"
Change the cache directory of ranger from \f(CW$XDG_CACHE_HOME\fR or ~/.cache/ranger to "dir".
.IP "\fB\-r\fR \fIdir\fR, \fB\-\-confdir\fR=\fIdir\fR" 14
.IX Item "-r dir, --confdir=dir"
Change the configuration directory of ranger from \f(CW$XDG_CONFIG_HOME\fR or ~/.config/ranger to "dir".
.IP \fB\-\-datadir\fR=\fIdir\fR 14
.IX Item "--datadir=dir"
Change the data directory of ranger from \f(CW$XDG_DATA_HOME\fR or ~/.local/share/ranger to "dir".
.IP \fB\-\-copy\-config\fR=\fIfile\fR 14
.IX Item "--copy-config=file"
Create copies of the default configuration files in your local configuration
directory.  Existing ones will not be overwritten.  Possible values: \fIall\fR,
\&\fIcommands\fR, \fIcommands_full\fR, \fIrc\fR, \fIrifle\fR, \fIscope\fR.
.Sp
Note: You can disable loading of the global \fIrc.conf\fR by exporting
\&\fIRANGER_LOAD_DEFAULT_RC=FALSE\fR in your environment. Only do so if your
configuration is complete, as settings that are not specified will be
undefined. See also: \fB\s-1FILES\s0\fR, \fB\s-1ENVIRONMENT\s0\fR
.Sp
\&\-\-copy\-config=\fBcommands\fR will copy only a small sample configuration file with
a thoroughly commented example.  It is recommended to keep this file tidy to
avoid getting defunct commands on ranger upgrades.  The full default
commands.py can be copied with \-\-copy\-config=\fBcommands_full\fR, but that file
will be ignored by ranger and serves only as a reference for making your own
commands.
.IP \fB\-\-choosefile\fR=\fItargetfile\fR 14
.IX Item "--choosefile=targetfile"
Allows you to pick a file with ranger.  This changes the behavior so that when
you open a file, ranger will exit and write the absolute path of that file into
\&\fItargetfile\fR.
.IP \fB\-\-choosefiles\fR=\fItargetfile\fR 14
.IX Item "--choosefiles=targetfile"
Allows you to pick multiple files with ranger.  This changes the behavior so
that when you open a file, ranger will exit and write the absolute paths of all
selected files into \fItargetfile\fR, adding one newline after each filename.
.IP \fB\-\-choosedir\fR=\fItargetfile\fR 14
.IX Item "--choosedir=targetfile"
Allows you to pick a directory with ranger.  When you exit ranger, it will
write the last visited directory into \fItargetfile\fR.
.IP \fB\-\-selectfile\fR=\fItargetfile\fR 14
.IX Item "--selectfile=targetfile"
Open ranger with \fItargetfile\fR selected. This is a legacy option, superseded by
the behavior for the POSITIONAL ARGUMENTS.
.IP \fB\-\-show\-only\-dirs\fR 14
.IX Item "--show-only-dirs"
Display only the directories. May be used in conjunction with
\&\fB\-\-choosedir\fR=\fItargetfile\fR.
.IP \fB\-\-list\-unused\-keys\fR 14
.IX Item "--list-unused-keys"
List common keys which are not bound to any action in the "browser" context.
This list is not complete, you can bind any key that is supported by curses:
use the key code returned by \f(CWgetch()\fR.
.IP \fB\-\-list\-tagged\-files\fR=\fItag\fR 14
.IX Item "--list-tagged-files=tag"
List all files which are tagged with the given tag.  Note: Tags are single
characters.  The default tag is "*"
.IP \fB\-\-profile\fR 14
.IX Item "--profile"
Print statistics of CPU usage on exit.
.IP \fB\-\-cmd\fR=\fIcommand\fR 14
.IX Item "--cmd=command"
Execute the command after the configuration has been read.  Use this option
multiple times to run multiple commands.
.IP \fB\-\-version\fR 14
.IX Item "--version"
Print the version and exit.
.IP "\fB\-h\fR, \fB\-\-help\fR" 14
.IX Item "-h, --help"
Print a list of options and exit.
.SH CONCEPTS
.IX Header "CONCEPTS"
This part explains how certain parts of ranger work and how they can be used
efficiently.
.SS TAGS
.IX Subsection "TAGS"
Tags are single characters which are displayed left of a filename.  You can use
tags however you want.  Press "t" to toggle tags and "ut" to remove any tags of
the selection. The default tag is an Asterisk ("*"), but you can use any tag by
typing \fI"<tagname>\fR.
.SS PREVIEWS
.IX Subsection "PREVIEWS"
By default, only text files are previewed, but you can enable external preview
scripts by setting the option \f(CW\*(C`use_preview_script\*(C'\fR and \f(CW\*(C`preview_files\*(C'\fR to true.
.PP
This default script is \fR\f(CI%rangerdir\fR\fI/data/scope.sh\fR. It contains more
documentation and calls to many external programs to generate previews. They
are automatically used when available but completely optional.
.IP "For general usage:" 2
.IX Item "For general usage:"
.RS 2
.PD 0
.IP \- 2
.PD
\&\f(CW\*(C`file\*(C'\fR for determining file types
.IP \- 2
\&\f(CW\*(C`chardet\*(C'\fR (Python package) for improved encoding detection of text files
.IP \- 2
\&\f(CW\*(C`sudo\*(C'\fR to use the "run as root" feature
.IP \- 2
\&\f(CW\*(C`python\-bidi\*(C'\fR (Python package) to display right-to-left file names correctly
(Hebrew, Arabic)
.RE
.RS 2
.RE
.IP "For enhanced file previews (with scope.sh):" 2
.IX Item "For enhanced file previews (with scope.sh):"
.RS 2
.PD 0
.IP \- 2
.PD
\&\f(CW\*(C`img2txt\*(C'\fR (from \f(CW\*(C`caca\-utils\*(C'\fR) for ASCII-art image previews
.IP \- 2
\&\f(CW\*(C`w3mimgdisplay\*(C'\fR, \f(CW\*(C`ueberzug\*(C'\fR, \f(CW\*(C`mpv\*(C'\fR, \f(CW\*(C`iTerm2\*(C'\fR, \f(CW\*(C`kitty\*(C'\fR (or other terminal
supporting the Kitty graphics protocol), \f(CW\*(C`terminology\*(C'\fR or \f(CW\*(C`urxvt\*(C'\fR for image
previews
.IP \- 2
\&\f(CW\*(C`convert\*(C'\fR (from \f(CW\*(C`imagemagick\*(C'\fR) to auto-rotate images and for image previews
.IP \- 2
\&\f(CW\*(C`rsvg\-convert\*(C'\fR (from \f(CW\*(C`librsvg\*(C'\fR) for SVG previews
.IP \- 2
\&\f(CW\*(C`ffmpegthumbnailer\*(C'\fR for video thumbnails
.IP \- 2
\&\f(CW\*(C`highlight\*(C'\fR, \f(CW\*(C`bat\*(C'\fR or \f(CW\*(C`pygmentize\*(C'\fR for syntax highlighting of code
.IP \- 2
\&\f(CW\*(C`atool\*(C'\fR, \f(CW\*(C`bsdtar\*(C'\fR, \f(CW\*(C`unrar\*(C'\fR and/or \f(CW\*(C`7zz\*(C'\fR to preview archives
.IP \- 2
\&\f(CW\*(C`bsdtar\*(C'\fR, \f(CW\*(C`tar\*(C'\fR, \f(CW\*(C`unrar\*(C'\fR, \f(CW\*(C`unzip\*(C'\fR and/or \f(CW\*(C`zipinfo\*(C'\fR (and \f(CW\*(C`sed\*(C'\fR) to preview
archives as their first image
.IP \- 2
\&\f(CW\*(C`lynx\*(C'\fR, \f(CW\*(C`w3m\*(C'\fR or \f(CW\*(C`elinks\*(C'\fR to preview html pages
.IP \- 2
\&\f(CW\*(C`pdftotext\*(C'\fR or \f(CW\*(C`mutool\*(C'\fR (and \f(CW\*(C`fmt\*(C'\fR) for textual pdf previews, \f(CW\*(C`pdftoppm\*(C'\fR to
preview as image
.IP \- 2
\&\f(CW\*(C`djvutxt\*(C'\fR for textual DjVu previews, \f(CW\*(C`ddjvu\*(C'\fR to preview as image
.IP \- 2
\&\f(CW\*(C`calibre\*(C'\fR or \f(CW\*(C`epub\-thumbnailer\*(C'\fR for image previews of ebooks
.IP \- 2
\&\f(CW\*(C`transmission\-show\*(C'\fR for viewing BitTorrent information
.IP \- 2
\&\f(CW\*(C`mediainfo\*(C'\fR or \f(CW\*(C`exiftool\*(C'\fR for viewing information about media files
.IP \- 2
\&\f(CW\*(C`odt2txt\*(C'\fR for OpenDocument text files (odt, ods, odp and sxw)
.IP \- 2
\&\f(CW\*(C`python\*(C'\fR or \f(CW\*(C`jq\*(C'\fR for JSON files
.IP \- 2
\&\f(CW\*(C`sqlite3\*(C'\fR for listing tables in SQLite database (and optionally \f(CW\*(C`sqlite\-utils\*(C'\fR for fancier box drawing.)
.IP \- 2
\&\f(CW\*(C`fontimage\*(C'\fR for font previews
.IP \- 2
\&\f(CW\*(C`openscad\*(C'\fR for 3D model previews (stl, off, dxf, scad, csg)
.IP \- 2
\&\f(CW\*(C`draw.io\*(C'\fR for draw.io <https://app.diagrams.net/> diagram previews (drawio
extension)
.RE
.RS 2
.RE
.PP
Install these programs (just the ones you need) and scope.sh will automatically
use them.
.PP
Independently of the preview script, there is a feature to preview images
by drawing them directly into the terminal. To enable this feature, set the
option \f(CW\*(C`preview_images\*(C'\fR to true and enable one of the image preview modes:
.PP
\fIiTerm2\fR
.IX Subsection "iTerm2"
.PP
This only works in iTerm2 compiled with image preview support, but works over
ssh.
.PP
To enable this feature, set the option \f(CW\*(C`preview_images_method\*(C'\fR to iterm2.
.PP
This feature relies on the dimensions of the terminal's font.  By default, a
width of 8 and height of 11 are used.  To use other values, set the options
\&\f(CW\*(C`iterm2_font_width\*(C'\fR and \f(CW\*(C`iterm2_font_height\*(C'\fR to the desired values.
.PP
iTerm's image protocol requires the image file to be split into chunks, the size
of which is set by the option \f(CW\*(C`iterm2_multipart_file_chunk_size\*(C'\fR.  It is set to
a sensible default, but can be tweaked for compatibility in certain situations.
Versions of \f(CW\*(C`tmux\*(C'\fR before 2.0 may need this set to 256.  Versions of iTerm
before 3.5.0 will need this set to 0, which reverts to iTerm's older image
protocol.
.PP
\fIsixel\fR
.IX Subsection "sixel"
.PP
This works in most terminal emulators, including all terminals based on VTE >=0.68
(e.g. GNOME Terminal), iTerm2, MinTTY, WezTerm and many others.
.PP
To enable this feature, install \f(CW\*(C`imagemagick\*(C'\fR and set the option
\&\f(CW\*(C`preview_images_method\*(C'\fR to sixel.
.PP
Preferred dithering method can be specified by setting \f(CW\*(C`sixel_dithering\*(C'\fR.
.PP
\fIkitty\fR
.IX Subsection "kitty"
.PP
This works in terminals supporting the Kitty graphics protocol. It requires PIL
(or pillow) to work.
Allows remote image previews, for example in an ssh session.
.PP
To enable this feature, set the option \f(CW\*(C`preview_images_method\*(C'\fR to kitty.
.PP
\fIterminology\fR
.IX Subsection "terminology"
.PP
This only works in terminology. It can render vector graphics, but works only locally.
.PP
To enable this feature, set the option \f(CW\*(C`preview_images_method\*(C'\fR to terminology.
.PP
\fIueberzug\fR
.IX Subsection "ueberzug"
.PP
Überzug is a command line utility which draws images on terminals using child
windows. It requires PIL (or pillow) and relies on X11. This makes it
compatible (in a limited way, i.e., tmux splits are not supported) with many
terminals and tmux but not the Linux console or Wayland.
.PP
To enable this feature, set the option \f(CW\*(C`preview_images_method\*(C'\fR to ueberzug.
.PP
\fIurxvt\fR
.IX Subsection "urxvt"
.PP
This only works in urxvt compiled with pixbuf support. Does not work over ssh.
.PP
Essentially this mode sets an image as a terminal background temporarily, so it
will break any previously set image background.
.PP
To enable this feature, set the option \f(CW\*(C`preview_images_method\*(C'\fR to urxvt.
.PP
\fIurxvt-full\fR
.IX Subsection "urxvt-full"
.PP
The same as urxvt but utilizing not only the preview pane but the whole terminal
window.
.PP
To enable this feature, set the option \f(CW\*(C`preview_images_method\*(C'\fR to urxvt-full.
.PP
\fIw3m\fR
.IX Subsection "w3m"
.PP
This does not work over ssh, requires certain terminals (tested on "xterm",
"urxvt", and "Alacritty") and is compatible with tmux and screen.
.PP
To enable this feature, install the program "w3m" and set the option
\&\f(CW\*(C`preview_images_method\*(C'\fR to w3m.
.PP
When using a terminal with a nonzero border which is not automatically detected, the w3m preview will be misaligned.
Use the \f(CW\*(C`w3m_offset\*(C'\fR option to manually adjust the image offset. This should be the same value as the terminal's border value.
.SS SELECTION
.IX Subsection "SELECTION"
The \fIselection\fR is defined as "All marked files IF THERE ARE ANY, otherwise
the current file."  Be aware of this when using the :delete command, which
deletes all files in the selection.
.PP
You can mark files by pressing <Space>, v, etc.  A yellow \fBMrk\fR symbol at the
bottom right indicates that there are marked files in this directory.
.SS MACROS
.IX Subsection "MACROS"
Macros can be used in commands to abbreviate things.
.PP
.Vb 6
\& %f   the highlighted file
\& %d   the path of the current directory
\& %s   the selected files in the current directory
\& %t   all tagged files in the current directory
\& %c   the full paths of the currently copied/cut files
\& %p   the full paths of selected files
.Ve
.PP
The macros \f(CW%f\fR, \f(CW%d\fR, \f(CW%p\fR, and \f(CW%s\fR also have upper case variants, \f(CW%F\fR, \f(CW%D\fR, \f(CW%P\fR, and
\&\f(CW%S\fR, which refer to the next tab.  To refer to specific tabs, add a number in
between.  (%7s = selection of the seventh tab.)
.PP
\&\f(CW%c\fR is the only macro which ranges out of the current directory. So you may
"abuse" the copying function for other purposes, like diffing two files which
are in different directories:
.PP
.Vb 2
\& Yank the file A (type yy), move to the file B, then type
\& @diff %c %f
.Ve
.PP
Macros for file paths are generally shell-escaped so they can be used in the
\&\f(CW\*(C`shell\*(C'\fR command.
.PP
When mapping keys you can use the placeholder <any>, the key entered in that
position can be used through the \f(CW%any\fR and \f(CW%any_path\fR macros. (When using
multiple <any> placeholders you can index the macros: \f(CW%any0\fR, \f(CW%any_path0\fR, \f(CW%any1\fR,
\&\f(CW%any_path1\fR...) The macro \f(CW%any\fR will be replaced with the key pressed in the
position of the <any> placeholder. The macro \f(CW%any_path\fR will be replaced with
the path of the bookmark mapped to the key pressed in the position of the
<any> placeholder. For example this macro can be used to echo the key that was
pressed after "c":
.PP
.Vb 1
\& map c<any> echo %any
.Ve
.PP
\&\f(CW%any\fR is used in the ranger configuration to create a keybinding for adding a
bookmark. c<set_bookmark> creates a bookmark for the current directory and the
key for the bookmark is the first supplied argument. In this case the key
pressed after "m":
.PP
.Vb 1
\& map m<any> set_bookmark %any
.Ve
.PP
The \f(CW%any_path\fR macro can be used to echo the path of the bookmark that is set to
the key pressed after "c":
.PP
.Vb 1
\& map c<any> echo %any_path
.Ve
.PP
A practical example of the use of \f(CW%any_path\fR is the pasting of cut/copied files
to a bookmarked directory:
.PP
.Vb 1
\& map p\*(Aq<any> paste dest=%any_path
.Ve
.PP
The macro \f(CW%rangerdir\fR expands to the directory of ranger's python library, you
can use it for something like this command:
 alias show_commands shell less \f(CW%rangerdir\fR/config/commands.py
.PP
\&\f(CW%confdir\fR expands to the directory given by \fB\-\-confdir\fR.
.PP
\&\f(CW%datadir\fR expands to the directory given by \fB\-\-datadir\fR.
.PP
The macro \f(CW%space\fR expands to a space character. You can use it to add spaces to
the end of a command when needed, while preventing editors to strip spaces off
the end of the line automatically.
.PP
To write a literal %, you need to escape it by writing %%.
.PP
Note that macros are expanded twice when using chain. For example, to insert
a space character in a chained command, you would write %%space:
 chain command1; command2%%space
.SS BOOKMARKS
.IX Subsection "BOOKMARKS"
Type \fBm<key>\fR to bookmark the current directory. You can re-enter this
directory by typing \fB`<key>\fR. <key> can be any letter or digit.  Unlike vim,
both lowercase and uppercase bookmarks are persistent.
.PP
Each time you jump to a bookmark, the special bookmark at key ` will be set
to the last directory. So typing "``" gets you back to where you were before.
.PP
Bookmarks are selectable when tabbing in the :cd command.
.PP
Note: The bookmarks ' (Apostrophe) and ` (Backtick) are the same.
.SS RIFLE
.IX Subsection "RIFLE"
Rifle is the file opener of ranger.  It can be used as a standalone program or
a python module.  It is located at \fR\f(CI$repo\fR\fI/ranger/ext/rifle.py\fR.  In contrast to
other, more simple file openers, rifle can automatically find installed
programs so it can be used effectively out of the box on a variety of systems.
.PP
It's configured in \fIrifle.conf\fR through a list of conditions and commands.
For each line the conditions are checked and if they are met, the respective
command is taken into consideration.  By default, simply the first matching
rule is used.  In ranger, you can list and choose rules by typing "r" or simply
by typing "<rulenumber><enter>".  If you use rifle standalone, you can list all
rules with the "\-l" option and pick a rule with "\-p <number>".
.PP
The rules, along with further documentation, are contained in
\&\fR\f(CI$repo\fR\fI/ranger/config/rifle.conf\fR.
.SS FLAGS
.IX Subsection "FLAGS"
Flags give you a way to modify the behavior of the spawned process.  They are
used in the commands \f(CW\*(C`:open_with\*(C'\fR (key "r") and \f(CW\*(C`:shell\*(C'\fR (key "!").
.PP
.Vb 6
\& f   Fork the process, i.e. run in background. Please use this flag
\&     instead of calling "disown" or "nohup", to avoid killing the
\&     background command when pressing Ctrl+C in ranger.
\& c   Run the current file only, instead of the selection
\& r   Run application with root privilege (requires sudo)
\& t   Run application in a new terminal window
.Ve
.PP
There are some additional flags that can currently be used only in the \f(CW\*(C`shell\*(C'\fR
command: (for example \f(CW\*(C`:shell \-w df\*(C'\fR)
.PP
.Vb 3
\& p   Redirect output to the pager
\& s   Silent mode.  Output will be discarded.
\& w   Wait for an Enter\-press when the process is done
.Ve
.PP
By default, all the flags are off unless otherwise specified in \fIrc.conf\fR key
bindings or \fIrifle.conf\fR rules. You can specify as many flags as you want. An
uppercase flag negates the effect: "ffcccFsf" is equivalent to "cs".
.PP
The terminal program name for the "t" flag is taken from the environment
variable \f(CW$TERMCMD\fR.  If it doesn't exist, it tries to extract it from \f(CW$TERM\fR,
uses "x\-terminal-emulator" as a fallback, and then "xterm" if that fails.
.PP
Examples: \f(CW\*(C`:open_with c\*(C'\fR will open the file that you currently point at, even
if you have selected other files.  \f(CW\*(C`:shell \-w df\*(C'\fR will run "df" and wait for
you to press Enter before switching back to ranger.
.SS PLUGINS
.IX Subsection "PLUGINS"
ranger's plugin system consists of python files which are located in
\&\fI~/.config/ranger/plugins/\fR and are imported in alphabetical order when
starting ranger.  A plugin changes rangers behavior by overwriting or extending
a function that ranger uses.  This allows you to change pretty much every part
of ranger, but there is no guarantee that things will continue to work in
future versions as the source code evolves.
.PP
Adding new commands via a plugin is as simple as specifying them like you would do
in the \fIcommands.py\fR.
.PP
There are some hooks that are specifically made for the use in plugins.  They
are functions that start with hook_ and can be found throughout the code.
.PP
.Vb 1
\& grep \*(Aqdef hook_\*(Aq \-r /path/to/rangers/source
.Ve
.PP
Also try:
.PP
.Vb 1
\& pydoc ranger.api
.Ve
.PP
Note that you should NOT simply overwrite a function unless you know what
you're doing.  Instead, save the existing function and call it from your new
one.  This way, multiple plugins can use the same hook.  There are several
sample plugins in the \fI/usr/share/doc/ranger/examples/\fR directory, including a
hello-world plugin that describes this procedure.
.SH "KEY BINDINGS"
.IX Header "KEY BINDINGS"
Key bindings are defined in the file \fR\f(CI%rangerdir\fR\fI/config/rc.conf\fR.  Check this
file for a list of all key bindings.  You can copy it to your local
configuration directory with the \-\-copy\-config=rc option.
.PP
Many key bindings take an additional numeric argument.  Type \fI5j\fR to move
down 5 lines, \fI2l\fR to open a file in mode 2, \fI10<Space>\fR to mark 10 files.
.PP
This list contains the most useful bindings:
.SS "MAIN BINDINGS"
.IX Subsection "MAIN BINDINGS"
.IP "h, j, k, l" 14
.IX Item "h, j, k, l"
Move left, down, up or right
.IP "^D or J, ^U or K" 14
.IX Item "^D or J, ^U or K"
Move a half page down, up
.IP "H, L" 14
.IX Item "H, L"
Move back and forward in the history
.IP gg 14
.IX Item "gg"
Move to the top
.IP G 14
.IX Item "G"
Move to the bottom
.IP "[, ]" 14
Move up and down in the parent directory.
.IP ^R 14
.IX Item "^R"
Reload everything
.IP F 14
.IX Item "F"
Toggle \fIfreeze_files\fR setting.  When active (indicated by a cyan \fIFROZEN\fR
message in the status bar), directories and files will not be loaded, improving
performance when all the files you need are already loaded.  This does not
affect file previews, which can be toggled with \fIzp\fR.  Also try disabling the
preview of directories with \fIzP\fR.
.IP ^L 14
.IX Item "^L"
Redraw the screen
.IP i 14
.IX Item "i"
Inspect the current file in a bigger window.
.IP E 14
.IX Item "E"
Edit the current file in \f(CW$VISUAL\fR otherwise \f(CW$EDITOR\fR otherwise "vim"
.IP S 14
.IX Item "S"
Open a shell in the current directory
.IP ? 14
Opens this man page
.IP W 14
.IX Item "W"
Opens the log window where you can review messages that pop up at the bottom.
.IP w 14
.IX Item "w"

CHANGELOG
This log documents changes between stable versions.

2015-10-04: version 1.7.2
* Fixed file name arguments passed to "sxiv" and "feh" when using ":flat"
* Fixed removal of empty directories when using ":rename"
* Fixed free disk space display on Mac OS X
* Fixed `examples/vim_file_chooser` to work with gvim too
* Fixed some other rare crashes and bugs
* Fixed downward mouse wheel scrolling
* Fixed warning about regex splits being drawin in title bar since python3.5
* Really fixed "S" key binding not working when SHELL=fish
* Improved doc/cheatsheet.svg
* Added some entries to rifle.conf
* Added key bindings pO and pP which work like po and pp but queue the operation
  in a first-in-first-out order.

2015-05-04: version 1.7.1
* Added doc/cheatsheet.svg
* Added examples/rc_emacs.conf, a config file adding emacs-like key bindings
* Added "env" keyword in rifle.conf
* Fixed ":bulkrename" command in python3
* Fixed "S" key binding not working when SHELL=fish

2015-04-13: version 1.7.0
* The default editor is now "vim" instead of "nano"
* Added automatic updates of tags when a file is renamed from within ranger
* Added "preview_images_method" which can be set to "iterm2" to use native
  iTerm2 image previews
* Added ":rename_append" command to rename files without the file extension
* Added ":linemode" command to change the way the files are displayed
  Try this out by pressing M followed by one of the suggested keys.
  New linemodes can be added with ranger.api.register_linemode().
* Added ":filter_inode_type" command to only show directories, files or links
* Added ":meta" command for managing custom file metadata
* Added ":flat" command for displaying subdirectories
* Added "solarized" colorscheme
* Added generic ability to use scope.sh for image previews
* Added video previews in scope.sh
* Added option "sort_unicode" to sort according to unicode, not ASCII
* ":mkdir" can now create multiple directory levels (like `mkdir -p`)
* ":help" (key binding "?") is now interactive
* ":find" (key binding "/") is now case insensitive by default
* "ranger --copy-config=all" now copies a short sample commands.py rather than
  the full one, so that you can update ranger without having broken commands.
  The full commands.py is still copied to ~/.config/ranger/commands_full.py.
* Fixed broken copying of symlinks

2013-05-24: Version 1.6.1
* Added support for version control systems, see:
  http://lists.nongnu.org/archive/html/ranger-users/2013-03/msg00007.html
* Added :scout command as a unified backend to :find, :search, etc
* Added "open_all_images" setting to remove the need for external scripts
  to handle opening of all images in a directory at once.
* Now previewing with "i" uses the whole available width.

2013-02-22: Version 1.6.0
* Overhauled all config files.  Please update them or use the --clean switch
* Added "examples/" directory to source code which contains sample programs or
  plugins that can be used together with ranger
* Added progress bars to copying, moving and directory loading processes
* Added feature to draw images inside the console using w3mimgdisplay (you need
  to add "set preview_images true" in rc.conf)
* Added a plugin system like in the program "anki", i.e. place any python file
  into ~/.config/ranger/plugins/ and it will be imported by ranger
* Added a separate file launcher named "rifle" that is configured through
  rifle.conf and is installed as a standalone program.
  Using "ranger [filename]" from the shell for opening files is deprecated now,
  please use "rifle [filename]" instead.
* Added "uq" keybinding to undo closed tabs
* Added :setlocal command to change settings for specific directories only
* Added :travel command to move more quickly to your destination
* Added 256 color support for scope.sh
* Added a real yes/no prompt for :delete command
* Added settings: confirm_on_delete, draw_progress_bar_in_status_bar,
  preview_images, status_bar_on_top, update_tmux_title
* Added commands: :mark_tag, :unmark_tag
* Added BSD-friendly setsid implementation
* Added as-you-type filtering for :filter command
* Replaced "options.py" file by :set commands in rc.conf
* Replaced "apps.py" file with rifle.conf
* Improved "r" key to interface with rifle
* Rewritten "scope.sh" in POSIX shell
* Changed copying/moving code to work without GNU coreutils
* Changed key to untag files from "T" to "ut"
* Changed the flag "d" (for detached) to "f" (for fork) in program launcher
* Changed appearance of keybinding-hints and bookmarks
* Changed tabs with 4 spaces in the source code (see PEP 8)
* Removed ranger.core.environment class
* Removed settings: colorscheme_overlay, draw_bookmark_borders, init_function,
  load_default_rc
* Fixed zombie process apocalypse
* Fixed draw_borders=true in combination with padding_right=false

2012-08-10: Version 1.5.5
* Ensure that detached programs continue to run when ranger is killed

2012-05-03: Version 1.5.4
* Added exiftool to scope.sh by default
* Fixed a crash when entering a directory with a unicode name
* Speedup in ranger.ext.get_executables

2012-03-05: Version 1.5.3
* Added --selectfile option that selects a certain file on startup
* Added --list-tagged-files option
* Added --cmd option to run commands on startup
* Added --profile option for additional debug information on exit
* Added a visual mode (activate with "V", deactivate with Esc)
* Added a reversed visual mode (activate with "uV")
* Added $RANGER_LEVEL environment variable which ranger sets to "1" or higher
  so programs can know that they were spawned from ranger
* Added run flag "r" for running with root privileges (needs sudo)
* Added run flag "t" for running in a new terminal (as specified in $TERMCMD)
* Added :relink command to change destinations of symlinks
* Added "dc" binding for getting the cumulative size of a directory
* Added "autoupdate_cumulative_size" option
* Added "pht" binding to Paste Hardlinked subTrees (like cp -l)

ranger.py
#!/usr/bin/python -O
# This file is part of ranger, the console file manager.  (coding: utf-8)
# License: GNU GPL version 3, see the file "AUTHORS" for details.

# =====================
# This embedded bash script can be executed by sourcing this file.
# It will cd to ranger's last location after you exit it.
# The first argument specifies the command to run ranger, the
# default is simply "ranger". (Not this file itself!)
# The other arguments are passed to ranger.
"""":
temp_file="$(mktemp -t "ranger_cd.XXXXXXXXXX")"
ranger="${1:-ranger}"
if [ -n "$1" ]; then
    shift
fi
"$ranger" --choosedir="$temp_file" "$@"
return_value="$?"
if chosen_dir="$(cat -- "$temp_file")" && [ -n "$chosen_dir" ] && [ "$chosen_dir" != "$PWD" ]; then
    cd -- "$chosen_dir"
fi
rm -f -- "$temp_file"
return "$return_value"
"""

from __future__ import (absolute_import, division, print_function)

import sys

# Need to find out whether or not the flag --clean was used ASAP,
# because --clean is supposed to disable bytecode compilation
ARGV = sys.argv[1:sys.argv.index('--')] if '--' in sys.argv else sys.argv[1:]
sys.dont_write_bytecode = '-c' in ARGV or '--clean' in ARGV

# Start ranger
import ranger  # NOQA pylint: disable=import-self,wrong-import-position
sys.exit(ranger.main())  # pylint: disable=no-member
